presto-main/src/main/java/com/facebook/presto/execution/QueryQueue.java

```diff
@@ -0,0 +1,94 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.concurrent.AsyncSemaphore;
+import org.weakref.jmx.Managed;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class QueryQueue
+{
+    private final int maxQueuedQueries;
+    private final AtomicInteger queryQueueSize = new AtomicInteger();
+    private final AsyncSemaphore<QueueEntry> asyncSemaphore;
+
+    QueryQueue(Executor queryExecutor, int maxQueuedQueries, int maxConcurrentQueries)
+    {
+        checkNotNull(queryExecutor, "queryExecutor is null");
+        checkArgument(maxQueuedQueries > 0, "maxQueuedQueries must be greater than zero");
+        checkArgument(maxConcurrentQueries > 0, "maxConcurrentQueries must be greater than zero");
+
+        this.maxQueuedQueries = maxQueuedQueries;
+        this.asyncSemaphore = new AsyncSemaphore<>(maxConcurrentQueries,
+                queryExecutor,
+                queueEntry -> {
+                    QueueableExecution queueableExecution = queueEntry.dequeue();
+                    if (queueableExecution != null) {
+                        queueableExecution.start();
+                        return queueableExecution.getCompletionFuture();
+                    }
+                    SettableFuture<?> settableFuture = SettableFuture.create();
+                    settableFuture.set(null);
+                    return settableFuture;
+                });
+    }
+
+    @Managed
+    public int getQueueSize()
+    {
+        return queryQueueSize.get();
+    }
+
+    public boolean enqueue(QueueableExecution queueableExecution)
+    {
+        if (queryQueueSize.incrementAndGet() > maxQueuedQueries) {
+            queryQueueSize.decrementAndGet();
+            return false;
+        }
+
+        // Add a callback to dequeue the entry if it is ever completed.
+        // This enables us to remove the entry sooner if is cancelled before starting,
+        // and has no effect if called after starting.
+        QueueEntry entry = new QueueEntry(queueableExecution);
+        queueableExecution.getCompletionFuture().addListener(entry::dequeue, MoreExecutors.directExecutor());
+        queueableExecution.getCompletionFuture().addListener(queryQueueSize::decrementAndGet, MoreExecutors.directExecutor());
```

Comment:
So this deviates from the previous behavior in how the queueing is accounted (not sure if that is what you were intending). Previously the element was considered dequeued (via the counter) as soon as it was pulled out and started/canceled. Now, the element is only considered dequeued when the task is finished/canceled.

Which means if you have a queue size of 1, nobody will be able to participate in the queue until the previous task completes.